File: spring-data-requery-java/src/main/java/com/coupang/springframework/data/requery/repository/support/RequeryEntityInformationSupport.java
package com.coupang.springframework.data.requery.repository.support;

import com.coupang.springframework.data.requery.core.RequeryOperations;
import com.coupang.springframework.data.requery.repository.query.DefaultRequeryEntityMetadata;
import com.coupang.springframework.data.requery.repository.query.RequeryEntityMetadata;
import com.coupang.springframework.data.requery.utils.EntityDataStoreUtils;
import io.requery.meta.EntityModel;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.repository.core.support.AbstractEntityInformation;
import org.springframework.util.Assert;

/**
 * RequeryEntityInformationSupport
 *
 * @author debop@coupang.com
 * @since 18. 6. 7
 */
@Slf4j
public abstract class RequeryEntityInformationSupport<T, ID>
    extends AbstractEntityInformation<T, ID> implements RequeryEntityInformation<T, ID> {

    private RequeryEntityMetadata<T> metadata;

    public RequeryEntityInformationSupport(Class<T> domainClass) {
        super(domainClass);
        this.metadata = new DefaultRequeryEntityMetadata(domainClass);
    }

    public String getEntityName() {
        return metadata.getEntityName();
    }

    public String getModelName() {
        return metadata.getModelName();
    }

    @SuppressWarnings({ "rawtypes", "unchecked" })
    public static <T> RequeryEntityInformation<T, ?> getEntityInformation(@NotNull Class<T> domainClass,
                                                                          @NotNull RequeryOperations operations) {
        Assert.notNull(domainClass, "Domain class must not be null.");
        Assert.notNull(operations, "RequeryOperations must not be null.");

        EntityModel entityModel = operations.getEntityModel();
        Assert.notNull(entityModel, "EntityModel must not be null!");
        log.debug("entityModel={}", entityModel);

        if (org.springframework.data.domain.Persistable.class.isAssignableFrom(domainClass)) {
            return new RequeryPersistableEntityInformation(domainClass, entityModel);
        } else {
            return new RequeryEntityModelEntityInformation<>(domainClass, entityModel);
        }
    }
}

File: spring-data-requery-kotlin/src/main/kotlin/org/springframework/data/requery/kotlin/repository/support/RequeryEntityInformationSupport.kt
package org.springframework.data.requery.kotlin.repository.support

import mu.KotlinLogging
import org.springframework.data.domain.Persistable
import org.springframework.data.repository.core.support.AbstractEntityInformation
import org.springframework.data.requery.kotlin.core.RequeryOperations
import org.springframework.data.requery.kotlin.repository.query.DefaultRequeryEntityMetadata
import kotlin.reflect.KClass
import kotlin.reflect.full.isSuperclassOf

/**
 * org.springframework.data.requery.repository.support.RequeryEntityInformationSupport
 *
 * @author debop
 */
abstract class RequeryEntityInformationSupport<E: Any, ID: Any>(override val kotlinType: KClass<E>)
    : AbstractEntityInformation<E, ID>(kotlinType.java), RequeryEntityInformation<E, ID> {

    companion object {
        private val log = KotlinLogging.logger { }

        @JvmStatic
        fun <E: Any, ID: Any> getEntityInformation(domainKlass: KClass<E>,
                                                   operations: RequeryOperations): RequeryEntityInformation<E, ID> {

            val entityModel = operations.entityModel

            log.debug { "domainClass=$domainKlass, entityModel=$entityModel" }


            //return if(Persistable::class.java.isAssignableFrom(kotlinType.java)) {
            return if(Persistable::class.isSuperclassOf(domainKlass)) {
                RequeryPersistableEntityInformation(domainKlass, entityModel)
            } else {
                RequeryEntityModelEntityInformation(domainKlass, entityModel)
            }
        }
    }

    private val metadata = DefaultRequeryEntityMetadata.of(kotlinType)

    override val entityName: String get() = metadata.entityName

    override val modelName: String get() = metadata.modelName

}